---
title: "GH LAB NGS Analysis"
output: html_notebook
---

```{r message=FALSE, warning=FALSE}
library(dplyr)
library(foreach)
library(ggplot2)


```


```{r}
data("mapping")
data("cazy")
data("nr100.genes")
data("metadata_ngs")
data("tax")


```

```{r}

head(mapping)

head(cazy)

head(nr100.genes)

head(metadata_ngs)


```



## CAZy

compute CAZy proportion by samples and do a PCA

```{r mapping}
nr100.genes$length = as.numeric(as.character(nr100.genes$length))
mapping.prop = prop.table(as.matrix(mapping)/nr100.genes$length,2)
rownames(mapping.prop) = nr100.genes$genes_id
```


```{r}
cazy_mapping = merge(cazy , mapping.prop, by.x="genes_id", by.y="row.names", all.x=TRUE)
cazy_summary = apply(cazy_mapping[,-c(1:6)], 2, tapply, cazy_mapping$Family, sum)
cazy.pca = dudi.pca(as.data.frame(t(log10(cazy_summary + 10^-8))), scannf=F, nf=3)
cazy_pca_metadata = merge(cazy.pca$li, metadata_ngs, by="row.names", all.x=TRUE)
```


plot PCA results and select cazy family having the highest contribution on component

```{r}
p_cazy_richness = ggplot(cazy_pca_metadata) + geom_text(aes(x=Axis1, y=Axis2, label=paste(Substrat, Souches, Time), col= genes.nb),size=4 )

idx1up   = which(cazy.pca$co[,1] > quantile(cazy.pca$co[,1], 0.975))
idx1down = which(cazy.pca$co[,1] < quantile(cazy.pca$co[,1], 0.225))

idx2up   = which(cazy.pca$co[,2] > quantile(cazy.pca$co[,2], 0.975))
idx2down = which(cazy.pca$co[,2] < quantile(cazy.pca$co[,2], 0.225))

cazy_comp = cazy.pca$co[unique(c(idx1up,idx1down,idx2up,idx2down)),]



p_cazy_family_pca = ggplot(cazy_comp) + geom_text(aes(x=Comp1, y=Comp2, label=row.names(cazy_comp)))

p_cazy_richness
p_cazy_family_pca

```
export figures in PDF

```{r}
pdf("figures/cazy_pca_v2.pdf", h=7,w=11)
p_cazy_richness
dev.off()

pdf("figures/cazy_family_pca_v2.pdf", h=7,w=11)
p_cazy_family_pca
dev.off()

```


```{r}
tmp = merge(melt(cazy_summary[c("GH13"),]) , metadata_ngs, all.x=TRUE, by="row.names")

p_GH13 = ggplot(tmp) + geom_bar(aes(paste(Substrat, Souches, Time, Row.names), y=value, fill=Substrat ), stat="identity")  + coord_flip() +
          ylab("prop GH13") + xlab("")


p_GH13_boxplot =ggplot(tmp) + geom_boxplot(aes(x=Substrat, y=value, fill=Substrat ))  + ylab("prop GH13") + xlab("")



tmp = merge(melt(cazy_summary[c("GH8"),]) , metadata_ngs, all.x=TRUE, by="row.names")

p_GH8 = ggplot(tmp) + geom_bar(aes(x=paste(Substrat, Souches, Time, Row.names), y=value, fill=Substrat ), stat="identity", position="dodge")  + coord_flip() +
          ylab("prop GH8") + xlab("")



p_GH8_boxplot = ggplot(tmp) + geom_boxplot(aes(x=Substrat, y=value, fill=Substrat ))  + ylab("prop GH8") + xlab("")


p_GH13_boxplot
p_GH8_boxplot

```

export figure in PDF

```{r message=FALSE, warning=FALSE}
pdf("figures/cazy_family_GH13_GH8_v2.pdf", h=7,w=15)
grid.arrange(p_GH13, p_GH8, ncol=2)
dev.off()




pdf("figures/cazy_family_GH13_GH8_boxplot.pdf", h=7,w=15)
grid.arrange(p_GH13_boxplot, p_GH8_boxplot, ncol=2)
dev.off()
```



## taxonomy

for each genes, a taxonomic association was done using BLAST

```{r}
#tax

idx = match(row.names(mapping.prop), tax$genes_id)

#mapping_tax = merge(mapping.prop, tax, all.x=TRUE, by.x="row.names", by.y="GeneName")


mapping_tax  = cbind(mapping.prop,tax[idx,])

mapping_tax  = cbind(mapping_tax, nr100_length_cdhit_genomes)

species = as.character(mapping_tax$species)

species[!(mapping_tax$genus %in% c("Lactobacillus"))] <- "others"



genus = as.character(mapping_tax$genus)

genus[!(mapping_tax$genus %in% c("Lactobacillus","Propionibacterium","Lactococcus","Staphylococcus","Bifidobacterium","Streptococcus","Pediococcus"))] <- "others"


mapping_tax = data.frame(mapping_tax, species_clean = species, genus_clean = genus)
```


```{r}
p_pectin_agx =
ggplot(mapping_tax) + geom_point(aes(x=Pec_M2_T08, y=AGX3_M2_T08), alpha=0.2) +
scale_y_log10("Pectin M2 T08") + scale_x_log10("AGX−3 M2 T08")



p_pectin_agx_density =
ggplot(mapping_tax) + stat_density2d(aes(x=Pec_M2_T08, y=AGX3_M2_T08, fill = ..level..), geom="polygon") +
scale_y_log10("Pectin M2 T08") + scale_x_log10("AGX−3 M2 T08")


p_pectin_agx_limits =
ggplot(mapping_tax) + geom_point(aes(x=Pec_M2_T08, y=AGX3_M2_T08), alpha=0.2) +
scale_y_log10("Pectin M2 T08",limits=c(0.00001, 0.001) ) + scale_x_log10("AGX−3 M2 T08",limits=c(0.00001, 0.001))


p_pectin_agx_genus =
ggplot(mapping_tax) + geom_point(aes(x=Pec_M2_T08, y=AGX3_M2_T08, color=genus_clean), alpha=0.2) +
scale_y_log10("Pectin M2 T08",limits=c(0.00001, 0.001) ) + scale_x_log10("AGX−3 M2 T08",limits=c(0.00001, 0.001)) +
guides(col = guide_legend(ncol = 2))



p_pectin_agx_species =
ggplot(mapping_tax) + geom_point(aes(x=Pec_M2_T08, y=AGX3_M2_T08, color=species_clean), alpha=0.2) +
scale_y_log10("Pectin M2 T08",limits=c(0.00001, 0.001) ) + scale_x_log10("AGX−3 M2 T08",limits=c(0.00001, 0.001)) +
guides(col = guide_legend(ncol = 2))




p_pectin_agx_genomes =
ggplot(mapping_tax) + geom_point(aes(x=Pec_M2_T08, y=AGX3_M2_T08, color=as.character(Lmu15)), alpha=0.2) +
scale_y_log10("Pectin M2 T08",limits=c(0.00001, 0.001) ) + scale_x_log10("AGX−3 M2 T08",limits=c(0.00001, 0.001)) +
guides(col = guide_legend(ncol = 2))



p_pectin_agx_genomes =
ggplot(mapping_tax) + geom_point(aes(x=Pec_M2_T08, y=AGX3_M2_T08, color=as.character(Lr47)), alpha=0.2) +
scale_y_log10("Pectin M2 T08",limits=c(0.00001, 0.001) ) + scale_x_log10("AGX−3 M2 T08",limits=c(0.00001, 0.001)) +
guides(col = guide_legend(ncol = 2))
```



```{r}
pdf("figures/pectin_agx_v2.pdf", h=7,w=12)
p_pectin_agx
p_pectin_agx_density
p_pectin_agx_limits
p_pectin_agx_genus
p_pectin_agx_species
dev.off()
```

## genomes


```{r eval=FALSE, include=FALSE}
#Genomes based analysis
library(doParallel)
registerDoParallel(cores=4)


genomes_names = names(mapping_tax[,55:140][,-6])


foreach( i = 1:length(genomes_names)) %dopar% {

library(ggplot2)  
  
g_name = genomes_names[i]


g = mapping_tax[,g_name] > 0

p_pectin_agx_genomes =
ggplot(cbind(mapping_tax,g), aes(x=Pec_M2_T08, y=AGX3_M2_T08) ) + geom_point(aes(color=g), alpha=0.2) +  scale_color_discrete(g_name) +
scale_y_log10("Pectin M2 T08",limits=c(0.00001, 0.001) ) + scale_x_log10("AGX−3 M2 T08",limits=c(0.00001, 0.001)) +
guides(col = guide_legend(ncol = 2))

#ggsave(paste0("figures/test/pectin_agx_v2_genomes_",g_name,".png"), p_pectin_agx_genomes)

png(paste0("figures/test/pectin_agx_v2_genomes_",g_name,".png"))

print(p_pectin_agx_genomes)

dev.off()


}
```


```{r eval=FALSE, include=FALSE}
genomes_names = names(mapping_tax[,55:140][,-6])

foreach( i = 60:length(genomes_names)) %dopar% {

  library(ggplot2)  
  
  
g_name = genomes_names[i]


g = mapping_tax[,g_name] > 0

p_pectin_agx_genomes =
ggplot(cbind(mapping_tax,g), aes(x=Pec_M2_T08, y=AGX3_M2_T08) ) + geom_point(aes(color=g), alpha=0.2) +  scale_color_discrete(g_name) +
scale_y_log10("Pectin M2 T08" ) + scale_x_log10("AGX−3 M2 T08") +
guides(col = guide_legend(ncol = 2))

#ggsave(paste0("figures/test/pectin_agx_v2_genomes_",g_name,".png"), p_pectin_agx_genomes)

png(paste0("figures/test2/pectin_agx_v2_genomes_",g_name,".png"))

print(p_pectin_agx_genomes)

dev.off()


}
```


select specific genes for each genomes

```{r}
#> table(A_sum>0, !is.na(mapping_tax$phylum))
#
#         FALSE   TRUE
#  FALSE  52786 361767
#  TRUE    5364 102020



## binarise relevant taxonomic information

mapping_tax$Pediococcus = 0
mapping_tax$Pediococcus[which(mapping_tax$genus=="Pediococcus")] <-1


mapping_tax$Leuconostoc = 0
mapping_tax$Leuconostoc[which(mapping_tax$genus=="Leuconostoc")] <-1


mapping_tax$Ljohnsonii = 0
mapping_tax$Ljohnsonii[which(mapping_tax$species=="Lactobacillus johnsonii")] <-1


mapping_tax$Lbrevis = 0
mapping_tax$Lbrevis[which(mapping_tax$species=="Lactobacillus brevis")] <-1


mapping_tax$Lfermentum = 0
mapping_tax$Lfermentum[which(mapping_tax$species=="Lactobacillus fermentum")] <-1

mapping_tax$Lgasseri = 0
mapping_tax$Lgasseri[which(mapping_tax$species=="Lactobacillus gasseri")] <-1


mapping_tax$Ldelbrueckii = 0
mapping_tax$Ldelbrueckii[which(mapping_tax$species=="Lactobacillus delbrueckii")] <-1



mapping_tax$genomes_prev = apply(mapping_tax[, c(53:147)] %>% select(-species_clean, -genus_clean, -GHlab), 1, sum)



#sum(table(mapping_tax$genus[mapping_tax$genomes_prev > 0 & mapping_tax$genomes_prev < 10]))

#mapping_tax$genomes_prev[mapping_tax$Leuconostoc==1]



sort(apply(mapping_tax[mapping_tax$genomes_prev == 1,c(55:147)] %>% select(-species_clean, -genus_clean, -GHlab), 2, sum))

par(mfrow=c(1,2))

hist(log10(mapping_tax$AGX3_M2_T08[which(mapping_tax$Lr47 !=0 &mapping_tax$genomes_prev == 1 )]), main="specific genes Lr47")
hist(log10(mapping_tax$AGX3_M2_T08[which(mapping_tax$Lr47 !=0  )]), main="all genes Lr47")


## summarize gene mapping by specific genes

mapping_tax_specific = mapping_tax[mapping_tax$genomes_prev == 1,]


genomes_spec = mapping_tax_specific[,c(53:147)]%>% select(-species_clean, -genus_clean, -GHlab)


mapping_tax_specific$genomes_tax =  names(genomes_spec)[apply(genomes_spec  , 1, function(x){ which(x != 0)  } )]



samples_genomes_abund = apply(mapping_tax_specific[1:35] , 2, tapply, mapping_tax_specific$genomes_tax, median)

#genomes.pca = dudi.pca(as.data.frame(t(samples_genomes_abund)) , scannf=F, nf=3)

genomes.pca = dudi.pca(as.data.frame(t(samples_genomes_abund[,-c(5,14,17,28)])) , scannf=F, nf=3)

scatter(genomes.pca)


p_amy_agx_genomes =
ggplot(mapping_tax) + geom_point(aes(x=Amy_M2_T08, y=AGX3_M2_T08, color=as.character(genus_clean)), alpha=0.2) +
scale_y_log10("Amy M2 T08" ) + scale_x_log10("AGX−3 M2 T08") #+
#guides(col = guide_legend(ncol = 2))

#genomes.pca = dudi.pca(as.data.frame(t(log10(samples_genomes_abund[,-c(5,14,17,28)] + 10^-8 ))) , scannf=F, nf=3)


#X= "Amy_M2_T08"

#Y= "AGX3_M2_T08"
```


function to plot specific genes

```{r}
specific_genes_plot = function(X,Y, check.noise=TRUE, noise_cut=10^-6, noise_input=10^-8, limits=c(5*10^-7,5*10^-4), color=TRUE ) {

  tmp   = data.frame(X = mapping_tax_specific[,X], Y=mapping_tax_specific[,Y], genomes_tax=mapping_tax_specific$genomes_tax)
  tmp_2 = data.frame(X = samples_genomes_abund[,X], Y = samples_genomes_abund[,Y], genomes_label= rownames(samples_genomes_abund))

  if(check.noise) {

    X_idx = which(tmp$X < noise_cut)
    Y_idx = which(tmp$Y < noise_cut)

    ni = log10(noise_input)

    tmp$X[X_idx] = 10^(rnorm(length(X_idx),ni,0.1))
    tmp$Y[Y_idx] = 10^(rnorm(length(Y_idx),ni,0.1))

    tmp_2$X[tmp_2$X < noise_cut] = noise_input
    tmp_2$Y[tmp_2$Y < noise_cut] = noise_input

  }

if(color){
  p_specific_genes = ggplot(tmp) + geom_point(aes(x=X, y=Y, color=genomes_tax), alpha=0.2) +
  scale_x_log10(X, limits=limits) + scale_y_log10(Y, limits=limits) +
  guides(col = guide_legend(ncol = 4)) + geom_text(data=tmp_2, aes(x=X,y=Y, label=genomes_label )) } else {

  p_specific_genes = ggplot(tmp) + geom_point(aes(x=X, y=Y), alpha=0.2) +
  scale_x_log10(X, limits=limits) + scale_y_log10(Y, limits=limits)

  }


  print(p_specific_genes)

}
```

genererate plot and export PDF

```{r}
#specific_genes_plot(X,Y)

pdf("figures/compare_replicat_noise.pdf", h=7, w=12)

specific_genes_plot("C_M2_T00","D_M2_T00",   check.noise=FALSE, limits=c(10^-7,5*10^-4), color=FALSE )
specific_genes_plot("C_M2_T00","D_M2_T00",      noise_cut=10^-6, noise_input=8*10^-7, limits=c(10^-7,5*10^-4), color=FALSE )



specific_genes_plot("A_M1_T00","B_M1_T00",   check.noise=FALSE, limits=c(10^-7,5*10^-4), color=FALSE )
specific_genes_plot("A_M1_T00","B_M1_T00",      noise_cut=10^-6, noise_input=8*10^-7, limits=c(10^-7,5*10^-4), color=FALSE )


dev.off()




pdf("figures/specific_genes_plot.pdf", h=7, w=12)

#specific_genes_plot("La_M1_T06","AGX3_M2_T05",   noise_cut=10^-6, noise_input=8*10^-7)
specific_genes_plot("AGX3_M2_T08","AGX2_M2_T08", noise_cut=10^-6, noise_input=8*10^-7)
specific_genes_plot("AGX1_M2_T08","AGX3_M2_T08", noise_cut=10^-6, noise_input=8*10^-7)
#specific_genes_plot("C_M2_T00","D_M2_T00",       noise_cut=10^-6, noise_input=8*10^-7)
specific_genes_plot("La_M1_T06","La_M2_T06",     noise_cut=10^-6, noise_input=8*10^-7)
specific_genes_plot("bglu_M1_T06","bglu_M2_T06",  noise_cut=10^-6, noise_input=8*10^-7)



dev.off()


pdf("figures/dynamic_genes_plot.pdf", h=7, w=12)

specific_genes_plot("C_M2_T00","D_M2_T00",      noise_cut=10^-6, noise_input=8*10^-7)
specific_genes_plot("AGX3_M2_T02","ARS_M2_T02", noise_cut=10^-6, noise_input=8*10^-7)
specific_genes_plot("AGX3_M2_T05","ARS_M2_T05", noise_cut=10^-6, noise_input=8*10^-7)
specific_genes_plot("AGX3_M2_T08","ARS_M2_T08", noise_cut=10^-6, noise_input=8*10^-7)


dev.off()


```

an example of scatter plot with 4 strains selected using two substrates
```{r message=FALSE, warning=FALSE}

mapping_tax_specific %>% select("ARS_M2_T08", "AGX3_M2_T08", genomes_tax) %>%
  filter(genomes_tax %in% c("Lpp230","Bbr32", "Lr108","Lmu15")) %>%
  #slice_sample(n=1000) %>%
  ggplot() + geom_point(aes(y=ARS_M2_T08, x=AGX3_M2_T08, color=genomes_tax), alpha=0.2) +
  scale_y_log10(limits=c(5*10^-8,5*10^-4)) +
  scale_x_log10(limits=c(5*10^-8,5*10^-4))

```


## CAZy vs Genomes analysis


```{r}





genomes_df = data.frame(genes_id=mapping_tax$genes_id, mapping_tax[,c(55:59,61:140)])

genomes_cazy = merge(cazy,genomes_df, by="genes_id", all.x=TRUE)
genomes_cazy = na.omit(apply(genomes_cazy[,7:91] %>% select(-species_clean, -genus_clean, -GHlab), 2, tapply, genomes_cazy$Family, sum))

genomes_cazy.pca = dudi.pca(as.data.frame(t(genomes_cazy)), scannf=F, nf=3)


samples_genomes.pca = dudi.pca(log10(samples_genomes_abund[row.names(genomes_cazy.pca$li),-c(5,14,17,28)]+10^-7), scannf=F, nf=3)



samples_cazy.coi =  coinertia(samples_genomes.pca, genomes_cazy.pca, scannf=F, nf=3)

plot(samples_cazy.coi)
randtest(samples_cazy.coi)





```

## strain dynamic over time by substrat

select strain having at least a rel. abundance > 10^-5 in one condition

```{r}


samples_genomes_abund_select =  samples_genomes_abund[apply(samples_genomes_abund, 1, max) > 10^-5,]

samples_genomes_metadata =  merge(t(samples_genomes_abund_select), genes.nb.metadata, by="row.names")

samples_genomes_metadata_melt = melt(samples_genomes_metadata, id.vars=c("genes.nb","Substrat","Souches","Time","Row.names"))

ggplot(samples_genomes_metadata_melt) + geom_point(aes(x=Time, y=value, color=variable, shape=Souches)) +
facet_grid(variable~Substrat, scale="free_y") + theme_bw() + theme(axis.text.x=element_text(size=5),  axis.text.y=element_text(size=5) )

pdf("figures/genomes_substrat_dynamics.pdf", w=10, h=8)
ggplot(samples_genomes_metadata_melt) + geom_point(aes(x=Time, y=value, color=variable, shape=Souches)) +
facet_grid(variable~Substrat, scale="free_y") + theme_bw() + theme(axis.text.x=element_text(size=5),  axis.text.y=element_text(size=5) )

dev.off()



```


## Pangenomes global

```{r fig.height=6, fig.width=16}
idx = apply(genomes_df[,-1] %>% select(-species_clean , -genus_clean),1,sum) > 0

pangenomes = genomes_df[idx, ]


sp2.lr.nr100    = vegan::specaccum(t(pangenomes[,grep("Lr",  colnames(pangenomes))]), "random")
sp2.lpp.nr100   = vegan::specaccum(t(pangenomes[,grep("Lpp", colnames(pangenomes))]), "random")
```


```{r fig.height=6, fig.width=16}
par(mfrow=c(1,2))

boxplot(t(sp2.lr.nr100$perm),col="yellow",  pch="+", ylab="Number of genes", xlab="number of genomes", main="lr NR100" )

boxplot(t(sp2.lpp.nr100$perm),col="lightblue",  pch="+", ylab="Number of genes", xlab="number of genomes", main="lpp NR100" )


pdf("figures/pangenomes_NR100.pdf", w=16, h=6)

par(mfrow=c(1,2))

boxplot(t(sp2.lr.nr100$perm),col="yellow",  pch="+", ylab="Number of genes", xlab="number of genomes", main="lr NR100" )

boxplot(t(sp2.lpp.nr100$perm),col="lightblue",  pch="+", ylab="Number of genes", xlab="number of genomes", main="lpp NR100" )

dev.off()
```



